File: backend/routers/workout_sessions.py
from fastapi import APIRouter, Depends, HTTPException
from sqlmodel import Session, select
from database import get_session
from auth import get_current_user
from datetime import date as Date
from models import WorkoutSession, WorkoutSessionCreate, WorkoutSessionRead, SessionExercise, Exercise, ExerciseWorkoutTypeLink, WorkoutType

router = APIRouter()


@router.post("/workout-sessions", response_model=WorkoutSessionRead)
def create_workout_session(workout_session: WorkoutSessionCreate, session: Session = Depends(get_session), user_id: str = Depends(get_current_user)):
    name = workout_session.name

    # If a template is chosen and no name is given, inherit the template name
    if workout_session.template_id and not name:
        template = session.get(WorkoutType, workout_session.template_id)
        if template and template.user_id == user_id:
            name = template.name

    item = WorkoutSession(date=workout_session.date, name=name, user_id=user_id)
    session.add(item)
    session.commit()
    session.refresh(item)

    if workout_session.template_id:
        exercises = session.exec(
            select(Exercise)
            .join(ExerciseWorkoutTypeLink)
            .where(ExerciseWorkoutTypeLink.workout_type_id == workout_session.template_id)
            .where(Exercise.user_id == user_id)
        ).all()
        for exercise in exercises:
            se = SessionExercise(workout_session_id=item.id, exercise_id=exercise.id)
            session.add(se)
        session.commit()
        session.refresh(item)

    return item


@router.get("/workout-sessions", response_model=list[WorkoutSessionRead])
def get_workout_sessions(date: Date = None, session: Session = Depends(get_session), user_id: str = Depends(get_current_user)):
    query = select(WorkoutSession).where(WorkoutSession.user_id == user_id)
    if date:
        query = query.where(WorkoutSession.date == date)
    items = session.exec(query).all()
    return items


@router.get("/workout-sessions/{workout_session_id}", response_model=WorkoutSessionRead)
def get_workout_session(workout_session_id: int, session: Session = Depends(get_session), user_id: str = Depends(get_current_user)):
    item = session.get(WorkoutSession, workout_session_id)
    if not item:
        raise HTTPException(status_code=404, detail="Session not found")
    if item.user_id != user_id:
        raise HTTPException(status_code=403, detail="Forbidden")
    return item


@router.delete("/workout-sessions/{workout_session_id}", status_code=204)
def delete_workout_session(workout_session_id: int, session: Session = Depends(get_session), user_id: str = Depends(get_current_user)):
    item = session.get(WorkoutSession, workout_session_id)
    if not item:
        raise HTTPException(status_code=404, detail="Session not found")
    if item.user_id != user_id:
        raise HTTPException(status_code=403, detail="Forbidden")
    session.delete(item)
    session.commit()

File: frontend/src/api/workoutSessions.ts
import client from "./client";
import type { ExerciseSet, SessionExercise, WorkoutSession } from "../types";

export const getWorkoutSessions = async (date?: string): Promise<WorkoutSession[]> => {
    const params = date ? { date } : {};
    const response = await client.get("/workout-sessions", { params });
    return response.data;
};

export const getWorkoutSessionById = async (workout_session_id: number): Promise<WorkoutSession> => {
    const response = await client.get(`/workout-sessions/${workout_session_id}`);
    return response.data;
};

export const addWorkoutSession = async (date: string, name?: string, template_id?: number): Promise<WorkoutSession> => {
    const response = await client.post("/workout-sessions", { date, name, template_id });
    return response.data;
};

export const addExerciseToSession = async (
    workout_session_id: number,
    exercise_id: number
): Promise<SessionExercise> => {
    const response = await client.post(
        `/workout-sessions/${workout_session_id}/exercises?exercise_id=${exercise_id}`
    );
    return response.data;
};

export const addSetToSessionExercise = async (session_exercise_id: number, exerciseSets: Partial<ExerciseSet[]>): Promise<ExerciseSet[]> => {
    const response = await client.post(`/session-exercise/${session_exercise_id}/sets`, exerciseSets);
    return response.data;
};

export const patchSet = async (set_id: number, patch: Partial<Pick<ExerciseSet, 'reps' | 'value' | 'unit'>>): Promise<void> => {
    await client.patch(`/sets/${set_id}`, patch);
};

export const deleteSet = async (set_id: number): Promise<void> => {
    await client.delete(`/sets/${set_id}`);
};

export const deleteWorkoutSession = async (id: number): Promise<void> => {
    await client.delete(`/workout-sessions/${id}`);
};

export const deleteSessionExercise = async (sessionExerciseId: number): Promise<void> => {
    await client.delete(`/session-exercises/${sessionExerciseId}`);
};
